### PR TITLE
Use jit interface thunkgen tech for more files

### PR DESCRIPTION
**Pull request by davidwrighton**
- Reduce the boilerplate logic when modifying the jit ee interface
 - 3 new files in superpmi are now autogenerated as well as 2 files in the jit directory are now autogenerated
- Reduce to only having one copy of the jit-ee interface guid

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/jit-contrib

File: src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
-//
-// Copyright (c) Microsoft. All rights reserved.
-// Licensed under the MIT license. See LICENSE file in the project root for full license information.
-//
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// DO NOT EDIT THIS FILE! It IS AUTOGENERATED
```
Comment (AndyAyersMS) [inline src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h:4]: Can you leave breadcrumbs here and in the other newly auto-generated files to hiht at how they can be regenerated?

**Review by AndyAyersMS — COMMENTED** [review]
Thanks! I had been thinking we should do something like this for a long time.

**Comment by AndyAyersMS** [comment]
@BruceForstall @sandreenko think you should look this over too.

**Comment by jkotas** [comment]
> Unfortunately requires compiling the libjitinterface with the PAL

Why is that? That's very unfortunate.

File: src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h:4]: Yes, please include pointers to the instructions for re-generation

File: src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs [hunk 2]
```diff
@@ -439,6 +448,271 @@ class JitInterfaceWrapper
...
+/**********************************************************************************/
+// clang-format off
+/**********************************************************************************/
+");
+
+            foreach (FunctionDecl decl in functionData)
+            {
+                tw.Write($"\n{decl.ReturnType.NativeTypeName} WrapICorJitInfo::{ decl.FunctionName}(");
+                bool isFirst = true;
+                foreach (Parameter param in decl.Parameters)
```
Comment (BruceForstall) [inline src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs:489]: nit? Looks like you could save lots of duplicated code by extracting a function to output function signatures, and function argument call lists

File: src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs [hunk 3]
```diff
@@ -480,15 +754,22 @@ static void Main(string[] args)
             else
             {
                 IEnumerable<FunctionDecl> functions = ParseInput(new StreamReader(args[0]));
-                using (TextWriter tw = new StreamWriter(args[1]))
-                {
-                    Console.WriteLine("Generating {0}", args[1]);
-                    WriteManagedThunkInterface(tw, functions);
-                }
-                using (TextWriter tw = new StreamWriter(args[2]))
+
```
Comment (BruceForstall) [inline src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs:757]: Seems like you should have at least minimal argument validation here, like checking the number of arguments, and a "usage" message that would also serve as a short description of what this tool is doing.

**Review by BruceForstall — COMMENTED** [review]
This should make updating the JIT-EE interface easier, which is much appreciated.

File: src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h:4]: https://github.com/dotnet/runtime/blob/master/docs/project/updating-jitinterface.md is the current doc if changes are needed there.

File: src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/gen.bat [hunk 4]
```diff
@@ -1,4 +1,4 @@
 pushd %~dp0
-call ..\..\..\..\..\..\..\dotnet.cmd run -- ThunkInput.txt ..\CorInfoBase.cs ..\..\..\aot\jitinterface\jitinterface.h
+call ..\..\..\..\..\..\..\dotnet.cmd run -- ThunkInput.txt ..\CorInfoBase.cs ..\..\..\aot\jitinterface\jitinterface.h ..\..\..\..\jit\ICorJitInfo_API_names.h ..\..\..\..\jit\ICorJitInfo_API_wrapper.hpp ..\..\..\..\ToolBox\superpmi\superpmi-shared\icorjitinfoimpl.h ..\..\..\..\ToolBox\superpmi\superpmi-shim-counter\icorjitinfo.cpp ..\..\..\..\ToolBox\superpmi\superpmi-shim-simple\icorjitinfo.cpp
```
Comment (sandreenko) [inline src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/gen.bat:2]: Nit: name these paths like `..\..\..\..\..\..\..\` etc?

File: src/coreclr/src/tools/aot/jitinterface/corinfoexception.h [hunk 5]
```diff
@@ -9,17 +9,17 @@ typedef char16_t WCHAR;
 typedef wchar_t WCHAR;
 #endif
 
-class CorInfoException
```
Comment (sandreenko) [inline src/coreclr/src/tools/aot/jitinterface/corinfoexception.h:12]: question: why was it renamed?

**Review by sandreenko — COMMENTED** [review]
looks good, I like the simplified jit ee interface changes.

File: src/coreclr/src/tools/aot/jitinterface/corinfoexception.h (hunk 5 above)
Comment (davidwrighton) [reply src/coreclr/src/tools/aot/jitinterface/corinfoexception.h:12]: CorInfoException is an enum type in corinfo.h

**Comment by davidwrighton** [comment]
Updated with code review feedback. I've stepped back from requiring the PAL, as that was causing odd problems. I may address that in a future change.

File: src/coreclr/src/tools/aot/jitinterface/jitinterface.h [hunk 6]
```diff
@@ -191,1474 +195,1660 @@ class JitInterfaceWrapper
...
-    virtual void expandRawHandleIntrinsic(void* pResolvedToken, void* pResult)
-    {
-        CorInfoException* pException = nullptr;
-        _callbacks->expandRawHandleIntrinsic(_thisHandle, &pException, pResolvedToken, pResult);
-        if (pException != nullptr)
-            throw pException;
-    }
+    virtual unsigned int getMethodAttribs(
+          void* ftn)
+{
```
Comment (jkotas) [inline src/coreclr/src/tools/aot/jitinterface/jitinterface.h:201]: Nit: The indentation is off.

Comment (davidwrighton) [reply src/coreclr/src/tools/aot/jitinterface/jitinterface.h:201]: The generator is ugly enough as it is. Unless there is pushback, I'd prefer to leave the indentation.

File: src/coreclr/src/inc/jiteeversionguid.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (BruceForstall) [inline src/coreclr/src/inc/jiteeversionguid.h:1]: Doesn't this file also need a license notice?

File: src/coreclr/src/inc/jiteeversionguid.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+// NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE NOTE
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+constexpr GUID JITEEVersionIdentifier = { /* 0d235fe4-65a1-487a-8553-c845496da901 */
+    0x0d235fe4,
+    0x65a1,
+    0x487a,
+    {0x85, 0x53, 0xc8, 0x45, 0x49, 0x6d, 0xa9, 0x01}
+};
```
Comment (BruceForstall) [inline src/coreclr/src/inc/jiteeversionguid.h:28]: nit: missing trailing end-of-line character?

File: src/coreclr/src/inc/corinfo.h [hunk 9]
```diff
@@ -184,37 +184,9 @@ TODO: Talk about initializing strutures before use
...
-
-constexpr GUID JITEEVersionIdentifier = { /* 0d235fe4-65a1-487a-8553-c845496da901 */
-    0x0d235fe4,
-    0x65a1,
-    0x487a,
-    {0x85, 0x53, 0xc8, 0x45, 0x49, 0x6d, 0xa9, 0x01}
-};
-
+#include "jiteeversionguid.h"
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (BruceForstall) [inline src/coreclr/src/inc/corinfo.h:190]: Seems like the "END JITEEVersionIdentifier" comment needs to be in the jiteeversionguid.h file.

We probably should have big, unmiss-able, comments at the top of all the JIT-EE interface files pointing to the JIT-EE version, but that's somewhat unrelated; this system only works when code reviewers know to ask for an interface GUID change.

Maybe the jiteeversionguid.h file should also point to the `updating-jitinterface.md` documentation.

**Review by BruceForstall — APPROVED** [review]
A few comment-related comments

**Review by jkotas — APPROVED** [review]

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**